frame/contracts/src/exec.rs

```diff
@@ -797,8 +797,10 @@ where
 				)?;
 			}
 
-			// Every call or instantiate also optionally transferres balance.
-			self.initial_transfer()?;
+			// Every non delegate call or instantiate also optionally transfers the balance.
+			if self.top_frame().delegate_caller.is_none() {
+				self.initial_transfer()?;
+			}
```

Comment:
I think we should move that check into the `initial_transfer` function. Also the comment should be updated with the **why** (why no transfer on delegate).